### DISCUSSION TITLE
StyleCop.Analyzers

### DISCUSSION THREAD

**Discussion by GeorgeLeithead** [CommunityToolkit/Maui#506] [General]
Hi all,

If not already discussed, could I suggest that we add [StyleCop.Analyzers](https://www.nuget.org/packages/StyleCop.Analyzers) to the various projects to encourage rule compliance, like with .editorconfig? Some rules are always going to be contentious (such as "SA1101: Prefix local calls with this", which can be configured to alert 'none').

**Comment by pictos** [comment]
We had that discussion internally before, in the XCT era, and we didn't want to do that. If I'm not wrong, was because the `.editorconfig` provide us with all rules that we need and we don't want to scare any new contributor. This project has the goal to be the first project for people that want to contribute to open source, even the junior level

**Reply by GeorgeLeithead** [reply]
Speaking as a `new contributor`, I have to say that having in place additional analyzers, such as StyleCop, helps the contributor meet the standards expected by the project.  I personally fel fowl of it a couple times, and was asked in a review to [Please organize elements in this file according to the following StyleCop rules:](https://github.com/CommunityToolkit/Maui/pull/481#discussion_r914244066) but there are no StyleCop rules enforced!
Also, provided the rules are applied correctly (warning rather than error, or NONE is not applied), these can be a fantastc guide for best practices and good developer habits, that all developers (Senior and junior) can benefit from.

**Reply by pictos** [reply]
We've just been using the `editorconfig`, it's odd that you don't get the errors and warnings regarding our code standard.

**Reply by GeorgeLeithead** [reply]
I do get a lot of guidance from the `.editorconfig`, for example, `dotnet_style_require_accessibility_modifiers = omit_if_default:error` really threw me a lot of problems, especially with it conflicting 'Code Clean Up' where I have `Add accessibility modifiers`.  So, perhaps it's a conflict in other Extensions/configurations that are not covered?  Similarly, I prefer implicit types, and not var... a number of times, someone else changed it back again on me.  I also prefer to use discard, this was again removed and replaced by others.  Another which I have no idea, is 'Block body constructors' or 'expression body for constructors'.  I had it changed from one to the other, back again and changed again, and back again... I honestly have zero idea which should be used.  Not sure if `.editorconfig` can be used to enforce such things, but perhaps it needs strengthening?

**Reply by GeorgeLeithead** [reply]
Has anyone had an opportunity to look into this, or verify if it is or is not an issue?

**Reply by pictos** [reply]
I would say it's something on your end. None of the core members had this issues and you're the first contributor that raise it up. If it's possible, can you try to run our project using a clean IDE version? (on VS we have the SafeMode and PresentationMode that doesn't use any extension or custom configuration)

**Reply by bijington** [reply]
So I am hoping that as part of the GravatarImageSource PR I can play with the .editorconfig settings and see how to improve the experience. I suspect that you have a number of settings on your machine that are different from the defaults whereas our settings are mostly the defaults and are not explicitly defined

**Reply by GeorgeLeithead** [reply]
Almost certainly true that I have some setting or extension difference. The trouble is, that so will many, many prospective contributors. Be that personal preferences, company standards, extensions, settings, etc.

I think therefore, the only way to enforce specific standards for any project, they need to be part of THAT project. Otherwise it'll become painful (has been a bit for myself) for new contributors and possibly turn them off.

**Reply by bijington** [reply]
Yes I agree, hopefully I can turn on all the defaults in .editorconfig and we should be good but I will have a play in the week and try applying to the GravatarImageSource PR

**Reply by TheCodeTraveler** [reply]
Thanks Shaun + George!!

Just to ensure this link doesn't get lost in the previous conversation on the AvatarView PR, here is the StyleCop rules that I pointed George to in order to keep methods/properties/fields etc organized inside of each class: https://stackoverflow.com/a/310967/5953643

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
